### ISSUE TITLE
Chef gem install requires elevated privileges on Windows (since 12.0.0 onwards)

### ISSUE THREAD

**Issue by tknerr** [chef/chef#3125]
Installing Chef as a gem on Windows worked fine in 11.x:
```
W:\tmp\sample-toplevel-cookbook>gem install chef -v '11.18.6' --no-ri --no-rdoc
Successfully installed chef-11.18.6-x86-mingw32
1 gem installed
```

Starting with Chef 12.0.0 (up to latest 12.1.2) I always get the error below:
```
W:\tmp\sample-toplevel-cookbook>gem install chef -v '12.0.0' --no-ri --no-rdoc
Fetching: chef-12.0.0-x86-mingw32.gem (100%)
Temporarily enhancing PATH to include DevKit...
Building native extensions.  This could take a while...
ERROR:  Error installing chef:
        ERROR: Failed to build gem native extension.

    W:/tools/chefdk/embedded/bin/ruby.exe -rubygems W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/bin/rake RUBYARCHDIR=W:/home/.chefdk/gem/ruby/2.0.0/extensions/x86-mingw32/2.0.0/chef-12.0.0-x86-mingw32 RUBYLIBDIR=W:/home/.chefdk/gem/ruby/2.0.0/extensions/x86-mingw32/2.0.0/chef-12.0.0-x86-mingw32
checking for windmc... yes
checking for windres... yes
checking for gcc... yes
windmc chef-log.man
windres -i chef-log.rc -o resource.o
gcc -o chef-log.dll -shared resource.o
rake aborted!
Errno::EIO: Input/output error - RegCreateKeyEx
W:/home/.chefdk/gem/ruby/2.0.0/gems/win32-eventlog-0.6.1/lib/win32/eventlog.rb:224:in `add_event_source'
W:/home/.chefdk/gem/ruby/2.0.0/gems/chef-12.0.0-x86-mingw32/ext/win32-eventlog/Rakefile:44:in `block in <top (required)>'
Tasks: TOP => default => register
(See full trace by running task with --trace)

rake failed, exit code 1

Gem files will remain installed in W:/home/.chefdk/gem/ruby/2.0.0/gems/chef-12.0.0-x86-mingw32 for inspection.
Results logged to W:/home/.chefdk/gem/ruby/2.0.0/extensions/x86-mingw32/2.0.0/chef-12.0.0-x86-mingw32/gem_make.out
```

Any ideas where this is coming from?

**Comment by tknerr** [comment]
A bit more details:
```
W:\home\.chefdk\gem\ruby\2.0.0\gems\chef-12.1.2-x86-mingw32\ext\win32-eventlog>rake register --trace
checking for windmc... yes
checking for windres... yes
checking for gcc... yes
** Invoke register (first_time)
** Invoke chef-log.dll (first_time, not_needed)
** Invoke resource.o (first_time, not_needed)
** Invoke chef-log.rc (first_time, not_needed)
** Invoke chef-log.man (first_time, not_needed)
** Execute register
rake aborted!
Errno::EIO: Input/output error - RegCreateKeyEx
W:/home/.chefdk/gem/ruby/2.0.0/gems/win32-eventlog-0.6.1/lib/win32/eventlog.rb:224:in `add_event_source'
W:/home/.chefdk/gem/ruby/2.0.0/gems/chef-12.1.2-x86-mingw32/ext/win32-eventlog/Rakefile:44:in `block in <top (required)>'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/task.rb:240:in `call'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/task.rb:240:in `block in execute'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/task.rb:235:in `each'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/task.rb:235:in `execute'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/task.rb:179:in `block in invoke_with_call_chain'
W:/tools/chefdk/embedded/lib/ruby/2.0.0/monitor.rb:211:in `mon_synchronize'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/task.rb:172:in `invoke_with_call_chain'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/task.rb:165:in `invoke'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:150:in `invoke_task'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:106:in `block (2 levels) in top_level'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:106:in `each'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:106:in `block in top_level'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:115:in `run_with_threads'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:100:in `top_level'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:78:in `block in run'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:176:in `standard_exception_handling'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/lib/rake/application.rb:75:in `run'
W:/home/.chefdk/gem/ruby/2.0.0/gems/rake-10.4.2/bin/rake:33:in `<top (required)>'
W:/home/.chefdk/gem/ruby/2.0.0/bin/rake:23:in `load'
W:/home/.chefdk/gem/ruby/2.0.0/bin/rake:23:in `<main>'
Tasks: TOP => register
```

**Comment by jdmundrawala** [comment]
It's trying to install the event logging stuff. 
You probably don't have the privileges to do so?

Maybe we should make this a best effort somehow.

**Comment by tknerr** [comment]
I'm not running with elevated privileges here. Should I have to?

**Comment by jdmundrawala** [comment]
You do have to to if you want it to install the eventlog stuff. You'll probably have to hack up the rakefile if you don't want this stuff installed.

**Comment by tknerr** [comment]
Yes, I can verify that it works when running with elevated privileges. 

+1 for making that a best-effort, so non-admin users can install that gem too.

**Comment by tknerr** [comment]
@jdmundrawala any idea how I could hack up the Rakefile locally? Everytime I run `gem install chef -v "12.1.2"` the whole `W:\home\.chefdk\gem\ruby\2.0.0\gems\chef-12.1.2-x86-mingw32` directory gets replaced with it's pristine contents.

Is there a sane way to do this?

**Comment by jdmundrawala** [comment]
you can git clone:
Hack up Rakefile
bundle exec rake gem
gem install pkg/chef...-x86-mingw32.gem

**Comment by tknerr** [comment]
Thanks @jdmundrawala, let me give a try

**Comment by tknerr** [comment]
Okay, after adding the begin/rescue block below to the Rakefile the rest of the gem installation runs fine. Looks like this was the only place that required elevated privileges.
```
  begin
    Win32::EventLog.add_event_source(
      :source => "Application",
      :key_name => "Chef",
      :event_message_file => dll_file,
      :category_message_file => dll_file
    )
  rescue Errno::EIO => e
    puts "Skipping event log registration due to missing privileges: #{e}"
  end
```

**Comment by tknerr** [comment]
On a second thought: why is adding the event source during gem install required anyway?

If its really needed, isn't that something the omnibus installer should rather take care of? It wouldn't have any effect anyway, as the chef gem comes pre-bundled with chef-omnibus, right?

**Comment by jdmundrawala** [comment]
Yea, I'm not really sure why I added the install part to the Rakefile. I think it made developing it easier and it also allows people who gem install to get that functionality. It is true that the omnibus package will take care of installing this (only for chef-client, not chef-dk).

**Comment by smurawski** [comment]
@jdmundrawala @tknerr Probably added to the rakefile since that is more likely to be done with admin rights (which are required to create new event sources).  Once an event source is created, it can be written to and read from via a lower permission account.

**Comment by tknerr** [comment]
@jdmundrawala @smurawski I guess this can be closed, since #3126 is merged to master (12.3.x)

Since the first 12.3.0 rc release might be a bit further away, I also created PR #3203 which applies the same fix on the 12-2-stable branch. Mind reviewing that and leaving a comment?

**Comment by tknerr** [comment]
fixed via #3126 and released with 12.3.0
